packages/components/src/IDReader/readers/QRReader/Scanner.tsx

```diff
@@ -37,34 +40,55 @@ const Scanner = (props: ScannerProps) => {
   const scanner = useRef<QRScanner>()
   const videoElement = useRef<HTMLVideoElement>(null)
   const [qrOn, setQrOn] = useState(true)
-  const { onError, onScan } = props
+  const { onError, onScan, validator } = props
+  const onScanSuccess = useCallback(
+    (result: QRScanner.ScanResult) => {
+      if (result.data) {
+        try {
+          const data = JSON.parse(result.data)
+          const validationError = validator && validator(data)
+          if (validationError) {
+            onError('invalid', new Error(validationError))
+            return
+          }
+          onScan(data)
+        } catch (error) {
+          // log detailed error message to console for debugging
+          // eslint-disable-next-line no-console
+          console.error(error)
+          onError('parse', new Error('Invalid JSON format'))
+        }
+      }
+    },
+    [onScan, onError, validator]
+  )
+  const onScanError = useCallback(
+    (error: Error) => {
+      onError('parse', error)
+    },
+    [onError]
+  )
 
   useEffect(() => {
     const currentVideoElement = videoElement?.current
     if (currentVideoElement && !scanner.current) {
-      scanner.current = new QRScanner(
-        currentVideoElement,
-        (result) => {
-          if (result.data) {
-            // TODO: handle parse error
-            onScan(JSON.parse(result.data))
-          }
-        },
-        {
-          // TODO: improve error handling
-          onDecodeError: () => onError('parse'),
-          preferredCamera: 'environment',
-          highlightCodeOutline: true,
-          highlightScanRegion: false
-        }
-      )
+      // implementation does not support the deprecated constructor overloads
+      // but supports the current signature. TS is throwing error. Need to have a closer
+      // look why TS is not able to detect the correct signature
```

Comment:
What do you mean with this, to what signature does TypeScript guide you here? This is an okay 'hack' but didn't quite catch what you're working around here

Reply:
Yeah odd, I don't get an editor error either while tsc fails for me locally. Okay, get you now.